File: code/chap02/python/DNA-FASTA-V3/dna_base_count_ver_3.py
#!/usr/bin/python
#-----------------------------------------------------
# Version-3
# This is a DNA-Base-Count in PySpark.
# The goal is to show how "DNA-Base-Count" works.
#------------------------------------------------------
# Input Parameters:
#    argv[1]: String, input path
#-------------------------------------------------------
# @author Mahmoud Parsian
#-------------------------------------------------------
from __future__ import print_function

import sys
from operator import add

from pyspark.sql import SparkSession
from collections import defaultdict
from pyspark import StorageLevel

#-----------------------------------
#
# we get an iterator: which represents
# a single partition of source RDD
#
# This function creates a hash map of DNA Letters
# "z" denotes the total number of FASTA Records for a single partition
#
def process_FASTA_partition(iterator):
    # create an empty dictionary
    hashmap = defaultdict(int)

    for fasta_record in iterator:
        if (fasta_record.startswith(">")):
            hashmap["z"] += 1
        else:
            chars = fasta_record.lower()
            for c in chars:
                hashmap[c] += 1
#           end-for
#   end-for
    #print("hashmap=", hashmap)
#
    # Python 2.x
    key_value_list = [(k, v) for k, v in hashmap.iteritems()]

    # Python 3.x
    #key_value_list = [(k, v) for k, v in hashmap.items()]

    #print("key_value_list=", key_value_list)
    return  key_value_list
#
#end-def
#-----------------------------------


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: ", __file__, " <input-path>", file=sys.stderr)
        exit(-1)

    # create an instance of SparkSession object
    spark = SparkSession\
        .builder\
        .appName("dna_base_count_ver_3")\
        .getOrCreate()

    input_path = sys.argv[1]
    print("input_path : ", input_path)

    #recordsRDD = spark.read.text(inputPath).rdd.map(lambda r: r[0])
    recordsRDD = spark.sparkContext.textFile(input_path)
    #print("recordsRDD.count() : ", recordsRDD.count())
    #recordsAsList = recordsRDD.collect()
    #print("recordsAsList : ", recordsAsList)
    #numOfPartitions = recordsRDD.getNumPartitions()
    #print("numOfPartitions : ", numOfPartitions)

    # if you do not have enough RAM, then do the following
    # MEMORY_AND_DISK = StorageLevel(True, True, False, False, 1)
    #recordsRDD.persist(StorageLevel(True, True, False, False, 1))

    pairsRDD = recordsRDD.mapPartitions(process_FASTA_partition)
    print("pairsRDD : debug")
    #pairsAsList = pairsRDD.collect()
    #print("pairsRDD : ", pairsAsList)

    frequenciesRDD = pairsRDD.reduceByKey(lambda x, y: x+y)
    print("frequenciesRDD : debug")
    frequenciesAsList = frequenciesRDD.collect()
    print("frequenciesAsList : ", frequenciesAsList)

    spark.stop()

File: code/chap02/scala/src/main/scala/org/data/algorithms/spark/ch02/DNABaseCountVER3.scala
package org.data.algorithms.spark.ch02

import org.apache.spark.sql.SparkSession
import org.data.algorithms.spark.ch02.DNABaseCountFastq.{drop3Records, processFASTQPartition}

import scala.collection.mutable

/**
 *-----------------------------------------------------
 * Version-3
 * This is a DNA-Base-Count in Spark Using Scala.
 * The goal is to show how "DNA-Base-Count" works.
 *------------------------------------------------------
 * Input Parameters:
 *    String, input path
 *-------------------------------------------------------
 *
 * @author Biman Mandal
 *-------------------------------------------------------
 */
object DNABaseCountVER3 {

  /**
   *-----------------------------------
   *
   * we get an iterator: which represents
   * a single partition of source RDD
   *
   * This function creates a hash map of DNA Letters
   * "z" denotes the total number of FASTA Records for a single partition
   *
   */
  def processFASTQPartition(iterator: Iterator[String]): Iterator[(Char, Int)] = {
    // create an empty dictionary
    val hashMap = new mutable.HashMap[Char, Int]()

    iterator.foreach(fastqRecord => {
      hashMap.put('z', hashMap.getOrElse('z', 0) + 1)
      val chars = fastqRecord.toLowerCase
      for (c <- chars)
        hashMap.put(c, hashMap.getOrElse(c, 0) + 1)
    })

    hashMap.iterator
  }

  def main(args: Array[String]): Unit = {
    val spark =
      SparkSession.builder()
        .appName("DNABaseCountVER3")
        .master("local[*]")
        .getOrCreate()

    val inputPath = args(0)
    println(s"inputPath : $inputPath")

    val recordsRdd =
      spark.sparkContext.textFile(inputPath)
        .map(_.toLowerCase)
    println(s"recordsRdd : ${recordsRdd.collect().mkString("Array(", ", ", ")")}")

    // if you do not have enough RAM, then do the following
    // recordsRdd.persist(StorageLevel.MEMORY_AND_DISK)
    val pairsRdd = recordsRdd.mapPartitions(processFASTQPartition)
    println("pairsRdd : debug")

    val frequenciesRdd = pairsRdd.reduceByKey((x, y) => x + y)
    println("frequenciesRdd : debug")
    val frequenciesAsList = frequenciesRdd.collect()
    println("frequenciesAsList : ", frequenciesAsList.mkString("Array(", ", ", ")"))

    spark.stop()
  }

}
